pennylane/ftqc/qubit_graph.py

```diff
@@ -0,0 +1,817 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains the data structures to represent qubits using a hierarchical memory model
+and defines the API to access and manipulate these data structures.
+"""
+
+import uuid
+import warnings
+from typing import Any, Optional, Union
+
+import networkx as nx
+
+# Fail-safe for algorithms that traverse the nested graph structure
+# In other words, we should never expect more than this many layers of nested QubitGraphs
+MAX_TRAVERSAL_DEPTH = 100
+
+
+class QubitGraph:
+    """A class to represent a hierarchical qubit memory model as nested graphs of qubits.
+
+    A QubitGraph is a qubit that contains a graph of underlying qubits, where each underlying qubit
+    is itself a QubitGraph. This representation allows for nesting of lower-level qubits with
+    arbitrary depth to allow easy insertion of arbitrarily many levels of abstractions between
+    logical qubits and physical qubits.
+
+    Args:
+        graph (graph-like, optional): The graph structure to use for the QubitGraph's underlying
+            qubits. Inputting None (the default), leaves the underlying qubit graph in an
+            uninitialized state, in which case one of the graph-initialization methods may be used
+            to define the structure of the underlying qubit graph. QubitGraph expects an undirected
+            graph as input, specifically an instance of the ``networkx.Graph`` class, although other
+            networkx graphs and graph-like types are also permitted. An object is considered
+            "graph-like" if it has both a 'nodes' and an 'edges' attribute. Furthermore, QubitGraph
+            expects a "flat" (unnested) graph, and any graph-like objects nested within the graph
+            structure used as input are ignored.
+        id (Any, optional): An identifier for this QubitGraph object. The identifier is generally an
+            integer, string, or a tuple of integers and strings, but it may be any object. Inputting
+            None (the default), assigns a random universally unique identifier (uuid) to ``id``.
+
+    **Examples**
+
+    Create a QubitGraph (with id=0) using a 2x2 Cartesian grid to define the structure of its
+    underlying qubits:
+
+    >>> import networkx as nx
+    >>> from pennylane.ftqc import QubitGraph
+    >>> q = QubitGraph(nx.grid_graph((2,2)), id=0)
+    >>> q
+    QubitGraph<id=0, loc=[]>
+    >>> for child in q.children:
+    ...     print(child)
+    QubitGraph<id=(0, 0), loc=[0]>
+    QubitGraph<id=(0, 1), loc=[0]>
+    QubitGraph<id=(1, 0), loc=[0]>
+    QubitGraph<id=(1, 1), loc=[0]>
+
+    Using the QubitGraph object defined above as a starting point, it's possible to nest other
+    QubitGraph objects in any of the nodes of its underlying qubit graph:
+
+    >>> q[(0,0)] = QubitGraph(nx.grid_graph((2,)))
```

Comment:
Can we change the "advertised UI" for this to

```suggestion
    >>> q[(0,0)].init_graph(nx.grid_graph((2,)))
```

My reasoning for this is that while the above works fine for individually indexing into nodes via their labels, if you do:

```
for child in q0.children:
    child = QubitGraph(nx.grid_graph((2,)))
```
 nothing is updated (which I think makes sense, I'm not advocating changing the behaviour there). So it seems more versatile to promote using
 
 ```
for child in q0.children:
    child.init_graph(nx.grid_graph((2,)))
```

Reply:
That's a good find, thanks for pointing this out. I agree, we should advertise using `init_graph()` for nesting over assignment of a new QubitGraph.